geopmdpy/geopmdpy/exporter.py

```diff
@@ -0,0 +1,201 @@
+#!/usr/bin/env python3
+#  Copyright (c) 2015 - 2024 Intel Corporation
+#  SPDX-License-Identifier: BSD-3-Clause
+#
+
+import sys
+import os
+from time import sleep
+from argparse import ArgumentParser
+from . import pio, topo, stats, loop, session, __version_str__
+from prometheus_client import start_http_server, Gauge, Counter, Summary
+
+_STARTUP_SLEEP = 0.005
+
+class PrometheusExporter:
+    """Prometheus exporter of GEOPM telemetry using the geopm.stats.Collector to
+    summarize statistics
+
+    """
+    def __init__(self, stats_collector):
+        self._stats_collector = stats_collector
+        self._stats_collector.report_table()
+        sleep(_STARTUP_SLEEP) # Take two samples to enable derivative signals
+        self._metric_names, self._metric_data = self._stats_collector.report_table()
+        self._gauges = [None if type(metric_val) is str else _create_prom_metric(metric_name.replace('-', '_'), metric_name, 'Gauge')
+                        for metric_name, metric_val in zip(self._metric_names, self._metric_data)]
+        self._num_metric = 0
+        for metric_idx, gauge in enumerate(self._gauges):
+            if gauge is not None:
+                 gauge.set_function(lambda exporter=self, metric_idx=metric_idx: \
+                                    exporter._get_metric(metric_idx))
+                 self._num_metric += 1
+        self._num_fresh = self._num_metric
+        self._is_fresh = len(self._metric_names) * [True]
+
+    def run(self, period, port):
+        """Run the GEOPM Prometheus exporter with geopm.stats.Collector
+
+        """
+        _start_http_server(port)
+        for _ in loop.TimedLoop(period):
+            pio.read_batch()
+            self._stats_collector.update()
+
+    def _refresh(self):
+        """Refresh all values in report cache
+
+        """
+        _, self._metric_data = self._stats_collector.report_table()
+        self._stats_collector.reset()
+        self._num_fresh = self._num_metric
+        self._is_fresh = len(self._metric_names) * [True]
+
+    def _get_metric(self, metric_idx):
+        """Get metric from cached report and update cache if all metrics have
+        been sampled at least once.
```

Comment:
So the cache is *only* updated when everything has been sampled?  What is it that enforces that all metrics are sampled?

Reply:
The Prometheus server is expected to sample each metric that is registered at least once per update loop.

Reply:
> ...at least once per update loop

If the Prometheus server samples twice in an update loop, isn't there an issue here?  I'm thinking that on that second sample, self._refresh() was called for one metric (the one being sampled a second time) but not another one (where it was previously sampled only once during this update loop) that the data would be inconsistent.  Wouldn't this be a problem for any derivative based signals?